### PR TITLE
Add support for Qwen3.5 MoE

### PR DESCRIPTION
**Pull request by michaelroyzen**
### Summary

- Adds Liger Kernel optimizations for the Qwen3.5 MoE model family (`qwen3_5_moe` / `qwen3_5_moe_text`), targeting Transformers v5+
- Qwen3.5 MoE combines Qwen3 Next's hybrid GDN/attention architecture with Sparse MoE (shared + routed experts), so the implementation mirrors Qwen3 Next's Liger integration: Gemma-style RMSNorm (`LigerRMSNormForQwen3Next`), fused SwiGLU experts (`LigerExperts`), and fused linear cross-entropy loss

### Changes

**New file:**
- `src/liger_kernel/transformers/model/qwen3_5_moe.py` — `lce_forward` for `Qwen3_5MoeForCausalLM`, based on the Qwen3 Next version with the `load_balancing_loss_func` import updated to point to Qwen3.5 MoE's local definition

**Modified files:**
- `src/liger_kernel/transformers/monkey_patch.py` — `apply_liger_kernel_to_qwen3_5_moe` function (RMSNorm, SwiGLU experts, fused LCE; RoPE disabled) with instance patching for norm layers, shared expert, and routed experts; registered as `qwen3_5_moe` and `qwen3_5_moe_text` in `MODEL_TYPE_TO_APPLY_LIGER_FN`
- `src/liger_kernel/transformers/__init__.py` — Export `apply_liger_kernel_to_qwen3_5_moe` in `TYPE_CHECKING`, `__getattr__`, and `__all__`
- `test/utils.py` — `revert_liger_kernel_to_qwen3_5_moe` for test cleanup
- `test/convergence/fp32/test_mini_models.py` — Availability check, imports, and `MiniModelConfig` entry for `mini_qwen3_5_moe`
- `test/transformers/test_monkey_patch.py` — `is_qwen3_5_moe_available` helper and `test_apply_liger_kernel_to_instance_for_qwen3_5_moe` verifying all patches are applied correctly

### Test plan

- [x] `test_apply_liger_kernel_to_instance_for_qwen3_5_moe` passes (monkey patch instance patching)
- [x] `mini_qwen3_5_moe` convergence test passes (fp32 mini model)
- [x] Existing Qwen3 Next and Qwen3 MoE tests still pass (no regressions)

### CODE REVIEW

**Comment by michaelroyzen** [comment]
@shimizust @Tcc0403

**Comment by michaelroyzen** [comment]
Convergence test passes
Screenshot 2026-02-26 at 2 48 14 PM

File: src/liger_kernel/transformers/model/qwen3_5_moe.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            labels=labels,
+            shift_labels=shift_labels,
+            hidden_size=self.config.hidden_size,
+            **kwargs,
+        )
+        loss, _, token_accuracy, predicted_tokens = unpack_cross_entropy_result(result)
+    else:  # if in inference model materialize logits
+        logits = self.lm_head(kept_hidden_states)
+        if labels is not None or shift_labels is not None:
+            loss = self.loss_function(logits, labels, self.vocab_size, **kwargs)
```
Comment (Tcc0403) [inline src/liger_kernel/transformers/model/qwen3_5_moe.py]: `shift_labels` should be passed as well

File: src/liger_kernel/transformers/monkey_patch.py [hunk 2]
```diff
@@ -2652,13 +2652,17 @@ def apply_liger_kernel_to_qwen3_next(
                 f"Unsupported qwen3_next model type. `model` must be `Qwen3NextForCausalLM`, `Qwen3NextModel`. Got: {type(model)}"
             )
 
+        _patch_rms_norm_module_for_qwen3_next = partial(
+            _patch_rms_norm_module, offset=1.0, casting_mode="gemma", in_place=False
+        )
```
Comment (Tcc0403) [inline src/liger_kernel/transformers/monkey_patch.py:2657]: great catch

Comment (Tcc0403) [reply src/liger_kernel/transformers/monkey_patch.py:2657]: Since we modified qwen3next code, could you attach test results including it?

**Review by Tcc0403 — COMMENTED** [review]
@Mecoli1219 can you take a look?

File: src/liger_kernel/transformers/model/qwen3_5_moe.py (hunk 1 above)
Comment (michaelroyzen) [reply src/liger_kernel/transformers/model/qwen3_5_moe.py]: Done

File: src/liger_kernel/transformers/monkey_patch.py (hunk 2 above)
Comment (michaelroyzen) [reply src/liger_kernel/transformers/monkey_patch.py:2657]: Sure, will run and include it shortly

**Comment by michaelroyzen** [comment]
Screenshot 2026-02-27 at 1 46 37 PM
Screenshot 2026-02-27 at 1 47 22 PM


Confirming Qwen3-Next still passes

**Comment by michaelroyzen** [comment]
Are we ready to merge @Tcc0403 @Mecoli1219?

**Comment by michaelroyzen** [comment]
Thanks, just rebased and ran `make checkstyle` @Mecoli1219
